UnityProject/Assets/Scripts/Tilemaps/Utils/TileList.cs

```diff
@@ -1,156 +1,191 @@
+			var chunkPos = GetChunkPos(position);
+
+			if (objects.TryGetValue(chunkPos, out var objectsOut))
+			{
+				objectsOut.Remove(position, obj);
+
+				if (objectsOut.HasObjects() == false)
+				{
+					objects[chunkPos] = null;
+				}
+			}
 		}
-	}
 
-	public void Remove(Vector3Int position, RegisterTile obj = null)
-	{
-		if (_objects.TryGetValue(position, out var objectsOut))
+		public void ReorderObjects(Vector3Int position)
 		{
-			if (obj == null)
+			var offset = 0;
+			if (position.x % 2 != 0)
 			{
-				objectsOut.Clear();
+				offset = position.y % 2 != 0 ? 1 : 0;
 			}
 			else
 			{
-				objectsOut.Remove(obj);
+				offset = position.y % 2 != 0 ? 3 : 2;
+			}
+
+			var chunkPos = GetChunkPos(position);
+
+			var i = 0;
+			foreach (var register in objects[chunkPos].Get(position))
+			{
+				if (register.OrNull()?.CurrentsortingGroup == null) continue;
+
+				register.CurrentsortingGroup.sortingOrder = (i * 4) + offset;
+				i++;
 			}
 		}
-	}
 
-	public void ReorderObjects(Vector3Int position)
-	{
-		var offset = 0;
-		if (position.x % 2 != 0)
+		public bool HasObjects(Vector3Int localPosition)
 		{
-			offset = position.y % 2 != 0 ? 1 : 0;
+			var chunkPos = GetChunkPos(localPosition);
+
+			return objects.TryGetValue(chunkPos, out var chunk) && chunk.HasObjects(localPosition);
 		}
-		else
+
+		public List<RegisterTile> Get(Vector3Int position)
 		{
-			offset = position.y % 2 != 0 ? 3 : 2;
+			var chunkPos = GetChunkPos(position);
+
+			return objects.TryGetValue(chunkPos, out var objectsOut) ? objectsOut.Get(position) : EmptyList;
 		}
-		var i = 0;
-		foreach (var register in _objects[position])
+
+		public IEnumerable<RegisterTile> Get(Vector3Int position, ObjectType type)
 		{
-			if (register.OrNull()?.CurrentsortingGroup == null) continue;
-			register.CurrentsortingGroup.sortingOrder = (i * 4) + offset;
-			i++;
-		}
-	}
+			var chunkPos = GetChunkPos(position);
 
-	public bool HasObjects(Vector3Int localPosition)
-	{
-		return _objects.ContainsKey(localPosition) && _objects[localPosition].Count > 0;
-	}
-	public List<RegisterTile> Get(Vector3Int position)
-	{
-		return _objects.TryGetValue(position, out var objectsOut) ? objectsOut : emptyList;
-	}
+			if (objects.TryGetValue(chunkPos, out var objectsOut) == false)
+			{
+				return EmptyList;
+			}
 
-	public IEnumerable<RegisterTile> Get(Vector3Int position, ObjectType type)
-	{
-		if (_objects.TryGetValue(position, out var objectsOut) == false)
-		{
-			return emptyList;
+			var list = new List<RegisterTile>();
+			foreach (var x in objectsOut.Get(position))
+			{
+				if (x.ObjectType == type)
+				{
+					list.Add(x);
+				}
+			}
+
+			return list;
 		}
-		var list = new List<RegisterTile>();
-		foreach (var x in objectsOut)
+
+		public void InvokeOnObjects(IRegisterTileAction action, Vector3Int localPosition)
 		{
-			if (x.ObjectType == type) {
-				list.Add(x);
+			tempRegisterTiles.Clear();
+			tempRegisterTiles.AddRange(Get(localPosition));
```

Comment:
How thread safe  is this tempRegisterTiles?

Reply:
probably isnt safe

Reply:
might be worth just slapping a Lock around it?

Reply:
I havent changed any logic, just moved it about. Was working fine before

Reply:
ok cool